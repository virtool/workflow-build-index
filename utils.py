import asyncio
import gzip
import typing

import aiofiles
import virtool_core.history.db
import virtool_core.otus.utils

ISOLATE_KEYS = [
    "id",
    "source_type",
    "source_name",
    "default"
]

OTU_KEYS = [
    "name",
    "abbreviation",
    "schema"
]

RIGHTS = [
    "build",
    "modify",
    "modify_otu",
    "remove"
]

SEQUENCE_KEYS = [
    "accession",
    "definition",
    "host",
    "sequence"
]


async def get_patched_otus(db, data_path, manifest):
    coros = list()

    for patch_id, patch_version in manifest.items():
        coros.append(virtool_core.history.db.patch_to_version(
            db,
            data_path,
            patch_id,
            patch_version
        ))

    return [j[1] for j in await asyncio.tasks.gather(*coros)]


async def write_sequences_to_file(path, sequences):
    """
    Write a FASTA file based on a given `Iterable` containing sequence documents.
    Headers will contain the `_id` field of the document and the sequence text is from the `sequence` field.
    :param path: the path to write the file to
    :param sequences: the sequences to write to file
    """
    async with aiofiles.open(path, "w") as f:
        for sequence in sequences:
            sequence_id = sequence["_id"]
            sequence_data = sequence["sequence"]

            line = f">{sequence_id}\n{sequence_data}\n"
            await f.write(line)


def get_sequences_from_patched_otus(
        otus: typing.Iterable[dict],
        data_type: str, sequence_otu_map: dict
) -> typing.Generator[dict, None, None]:
    """
    Return sequence documents based on an `Iterable` of joined OTU documents. Writes a map of sequence IDs to OTU IDs
    into the passed `sequence_otu_map`.
    If `data_type` is `barcode`, all sequences are returned. Otherwise, only sequences of default isolates are returned.
    :param otus: an Iterable of joined OTU documents
    :param data_type: the data type of the parent reference for the OTUs
    :param sequence_otu_map: a dict to populate with sequence-OTU map information
    :return: a generator that yields sequence documents
    """
    for otu in otus:
        otu_id = otu["_id"]

        for isolate in otu["isolates"]:
            for sequence in isolate["sequences"]:
                sequence_id = sequence["_id"]
                sequence_otu_map[sequence_id] = otu_id

        if data_type == "barcode":
            for sequence in virtool_core.otus.utils.extract_sequences(otu):
                yield sequence
        else:
            for sequence in virtool_core.otus.utils.extract_default_sequences(otu):
                yield sequence


async def export(db, data_path, ref_id):
    otu_list = list()

    query = {
        "reference.id": ref_id,
        "last_indexed_version": {
            "$ne": None
        }
    }

    async for document in db.otus.find(query):
        _, joined, _ = await virtool_core.history.db.patch_to_version(
            db,
            data_path,
            document["_id"],
            document["last_indexed_version"]
        )

        otu_list.append(joined)

    return clean_export_list(otu_list)


def clean_export_list(otus):
    cleaned = list()

    otu_keys = OTU_KEYS + ["_id"]
    sequence_keys = SEQUENCE_KEYS + ["_id"]

    for otu in otus:
        try:
            otu["_id"] = otu["remote"]["id"]
        except KeyError:
            pass

        for isolate in otu["isolates"]:
            for sequence in isolate["sequences"]:
                try:
                    sequence["_id"] = sequence["remote"]["id"]
                except KeyError:
                    pass

        cleaned.append(clean_otu(otu, otu_keys, sequence_keys))

    return cleaned


def clean_otu(otu, otu_keys=None, sequence_keys=None):
    otu_keys = otu_keys or OTU_KEYS
    sequence_keys = sequence_keys or SEQUENCE_KEYS

    cleaned = {key: otu.get(key) for key in otu_keys}

    cleaned.update({
        "isolates": list(),
        "schema": otu.get("schema", list())
    })

    for isolate in otu["isolates"]:
        cleaned_isolate = {key: isolate[key] for key in ISOLATE_KEYS}
        cleaned_isolate["sequences"] = list()

        for sequence in isolate["sequences"]:
            cleaned_sequence = {key: sequence[key] for key in sequence_keys}

            for key in ["segment", "target"]:
                try:
                    cleaned_sequence[key] = sequence[key]
                except KeyError:
                    pass

            cleaned_isolate["sequences"].append(cleaned_sequence)

        cleaned["isolates"].append(cleaned_isolate)

    return cleaned


def compress_json_with_gzip(json_string: str, target: str):
    """
    Compress the JSON string to a gzipped file at `target`.
    """
    with gzip.open(target, 'wb') as f:
        f.write(bytes(json_string, "utf-8"))
