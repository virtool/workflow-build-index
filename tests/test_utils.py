import os
import types

import aiohttp.test_utils
import pytest

import utils


@pytest.fixture
def fake_otus():
    return [
        {
            "_id": "foo",
            "isolates": [
                {
                    "id": "foo_1",
                    "default": True,
                    "sequences": [
                        {
                            "_id": "1",
                            "sequence": "AGAGGATAGAGACACA"
                        },
                        {
                            "_id": "2",
                            "sequence": "GGGTAGTCGATCTGGC"
                        }
                    ]
                },
                {
                    "id": "foo_2",
                    "default": False,
                    "sequences": [
                        {
                            "_id": "3",
                            "sequence": "TTTAGAGTTGGATTAC",
                            "default": True
                        },
                        {
                            "_id": "4",
                            "sequence": "AAAGGAGAGAGAAACC",
                            "default": True
                        }
                    ]
                },
            ]
        },
        {
            "_id": "bar",
            "isolates": [
                {
                    "id": "bar_1",
                    "default": True,
                    "sequences": [
                        {
                            "_id": "5",
                            "sequence": "TTTGAGCCACACCCCC"
                        },
                        {
                            "_id": "6",
                            "sequence": "GCCCACCCATTAGAAC"
                        }
                    ]
                }
            ]
        }
    ]


@pytest.mark.parametrize("data_type", ["genome", "barcode"])
def test_get_sequences_from_patched_otus(data_type, data_regression, fake_otus):
    sequence_otu_dict = dict()

    sequences = utils.get_sequences_from_patched_otus(
        fake_otus,
        data_type,
        sequence_otu_dict
    )

    assert isinstance(sequences, types.GeneratorType)

    data_regression.check(list(sequences), basename=f"sequences_{data_type}")
    data_regression.check(sequence_otu_dict, basename=data_type)


@pytest.mark.asyncio
async def test_write_sequences_to_file(file_regression, tmpdir):
    sequences = [
        {
            "_id": "foo",
            "sequence": "ATTGAGAGATAGAGACAC"
        },
        {
            "_id": "bar",
            "sequence": "GGGTACGAGTTTCTATCG"
        },
        {
            "_id": "baz",
            "sequence": "GGCTTCGGACTTTTTTCG"
        }
    ]

    path = os.path.join(str(tmpdir), "output.fa")

    await utils.write_sequences_to_file(path, sequences)

    with open(path, "r") as f:
        file_regression.check(f.read())
