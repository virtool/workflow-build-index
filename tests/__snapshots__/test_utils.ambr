# serializer version: 1
# name: test_get_sequences_from_patched_otus[barcode][sequence_otu_dict]
  dict({
    '1': 'foo',
    '2': 'foo',
    '3': 'foo',
    '4': 'foo',
    '5': 'bar',
    '6': 'bar',
  })
# ---
# name: test_get_sequences_from_patched_otus[barcode][sequences]
  list([
    dict({
      '_id': '1',
      'sequence': 'AGAGGATAGAGACACA',
    }),
    dict({
      '_id': '2',
      'sequence': 'GGGTAGTCGATCTGGC',
    }),
    dict({
      '_id': '3',
      'default': True,
      'sequence': 'TTTAGAGTTGGATTAC',
    }),
    dict({
      '_id': '4',
      'default': True,
      'sequence': 'AAAGGAGAGAGAAACC',
    }),
    dict({
      '_id': '5',
      'sequence': 'TTTGAGCCACACCCCC',
    }),
    dict({
      '_id': '6',
      'sequence': 'GCCCACCCATTAGAAC',
    }),
  ])
# ---
# name: test_get_sequences_from_patched_otus[genome][sequence_otu_dict]
  dict({
    '1': 'foo',
    '2': 'foo',
    '3': 'foo',
    '4': 'foo',
    '5': 'bar',
    '6': 'bar',
  })
# ---
# name: test_get_sequences_from_patched_otus[genome][sequences]
  list([
    dict({
      '_id': '1',
      'sequence': 'AGAGGATAGAGACACA',
    }),
    dict({
      '_id': '2',
      'sequence': 'GGGTAGTCGATCTGGC',
    }),
    dict({
      '_id': '5',
      'sequence': 'TTTGAGCCACACCCCC',
    }),
    dict({
      '_id': '6',
      'sequence': 'GCCCACCCATTAGAAC',
    }),
  ])
# ---
